exec/java-exec/src/main/java/org/apache/drill/exec/store/parquet/AbstractParquetScanBatchCreator.java

```diff
@@ -68,76 +84,144 @@ protected ScanBatch getBatch(ExecutorFragmentContext context, AbstractParquetRow
+
+              // The file status for this file
+              FileStatus fileStatus = fs.getFileStatus(rowGroup.getPath());
+
+              // The file metadata (only for the columns used in the filter)
+              fileMetadataV4 = Metadata.getParquetFileMetadata_v4(tableMetadataV4, footer, fileStatus, fs, false, true, columnsInExpr, readerConfig);
+
+              prevRowGroupPath = rowGroup.getPath(); // for next time
+            }
+
+            MetadataBase.RowGroupMetadata rowGroupMetadata = fileMetadataV4.getFileMetadata().getRowGroups().get(rowGroup.getRowGroupIndex());
+
+            Map<SchemaPath, ColumnStatistics> columnsStatistics = ParquetTableMetadataUtils.getRowGroupColumnStatistics(tableMetadataV4, rowGroupMetadata);
 
-      } catch (IOException e) {
-        throw new ExecutionSetupException(e);
+            //
+            // Perform the Run-Time Pruning - i.e. Skip this rowgroup if the match fails
+            //
+            RowsMatch match = FilterEvaluatorUtils.matches(filterPredicate, columnsStatistics, footerRowCount);
+            if (timer != null) { // if tracing
+              long timeToRead = timer.elapsed(TimeUnit.MICROSECONDS);
+              logger.trace("Run-time pruning: {} row-group {} (RG index: {} row count: {}), took {} usec", match == RowsMatch.NONE ? "Excluded" : "Included", rowGroup.getPath(),
+                rowGroupIndex, footerRowCount, timeToRead);
+            }
+            if (match == RowsMatch.NONE) {
+              rowgroupsPruned++; // one more RG was pruned
+              if (firstRowGroup == null) {  // keep first RG, to be used in case all row groups are pruned
+                firstRowGroup = rowGroup;
+                firstFooter = footer;
+              }
+              continue; // This Row group does not comply with the filter - prune it out and check the next Row Group
+            }
+          }
+
+          mapWithMaxColumns = createReaderAndImplicitColumns(context, rowGroupScan, oContext, columnExplorer, readers, implicitColumns, mapWithMaxColumns, rowGroup,
+           fs, footer /*, false */);
       }
+
+      // in case all row groups were pruned out - create a single reader for the first one (so that the schema could be returned)
+      if ( readers.size() == 0 && firstRowGroup != null ) {
+        logger.trace("All row groups were pruned out. Returning the first: {} (row count {}) for its schema", firstRowGroup.getPath(), firstRowGroup.getNumRecordsToRead());
+        DrillFileSystem fs = fsManager.get(rowGroupScan.getFsConf(firstRowGroup), firstRowGroup.getPath());
+        mapWithMaxColumns = createReaderAndImplicitColumns(context, rowGroupScan, oContext, columnExplorer, readers, implicitColumns, mapWithMaxColumns, firstRowGroup, fs,
+          firstFooter /*, true */);
+      }
+
+      // Update stats (same in every reader - the others would just overwrite the stats)
+      for (RecordReader rr : readers ) {
+        if ( rr instanceof ParquetRecordReader ) {
```

Comment:
It would be useful to have this stats for the `DrillParquetReader` as well.

Reply:
Done in a followup commit